ApiData.ts

```diff
@@ -0,0 +1,264 @@
+function debounce<F extends (...args: any) => any>(func: F, waitFor: number) {
+	let timeout: number;
+
+	return (...args: Parameters<F>): Promise<ReturnType<F>> => {
+		clearTimeout(timeout);
+
+		return new Promise((resolve) => {
+			timeout = setTimeout(() => {
+				resolve(func(...args));
+			}, waitFor);
+		});
+	};
+}
+
+/** Constants for grid calculation
+* GRID_RATIO represents the ratio of the grid's height to the window's height.
+*/
+const GRID_RATIO = 9 / 20;
+const ROW_HEIGHT = 16;
+
+/** manage data and settings on the grid */
+class ApiData {
+
+	pageSize: number;
+	currentPage: number = 1;
+	data: GridData[] = [];
+	totalItems: number = 0;
+	columnNames: ColumnName[] = [];
+	maxGridHeight: number = 0;
+	firstVal: number = 0;
+	lastVal: number = -1;
+	maxRange: number;
+
+	constructor(pageSize: number) {
+		this.pageSize = pageSize;
+		this.maxRange = 0;
+	}
+
+	/** Initialize method to set up the grid */
+	initialize(): Promise<void> {
+		this.adjustGridHeight();
+		return this.recordCount()
+			.then(() => this.fetchColumns())
+			.then(() => this.fetchAndDisplayRecords())
+			.then(() => this.setupControls());
+	}
+
+	/** Fetch total record count from the server,fetches data from an API and populates class properties */
+	recordCount(): Promise<void> {
+		return this.fetchData('http://localhost:2050/recordCount')
+			.then((response: number | string) => {
+				const totalItems = <number>response;
+				this.totalItems = totalItems;
+			})
+			.catch(error => {
+				console.error('Failed to fetch record count:', error);
+				alert('Failed to fetch record count.');
+				throw error;
+			});
+	}
+
+	/** Use the fetchData() func to make an HTTP request to the API endpoint and process the data */
+	fetchColumns(): Promise<void> {
+		return this.fetchData('http://localhost:2050/columns')
+			.then((response: number | string) => {
+				const res = JSON.parse(<string>(response));
+				this.columnNames = res.map((columnName: string) => ({ name: columnName }));
+				// Initialize the 'data' property as an empty array of GridData objects
+				this.data = new Array<GridData>(this.columnNames.length);
+			})
+			.catch(() => {
+				throw ('Failed to fetch columns.');
+			});
+	}
+
+	/** Get records from API for fetch and search functionality */
+	fetchAndProcessRecords(from: number, to: number): Promise<GridData[]> {
+		$('#spinner').show();
+		$('#grid').hide();
+
+		return this.fetchData(`http://localhost:2050/records?from=${from}&to=${to}`)
+			.then((response: number | string) => {
+				const res = JSON.parse(<string>(response));
+				const processedData = res.map((record: string) => {
+					const obj: GridData = {};
+					for (let j = 0; j < this.columnNames.length && j < record.length; j++) {
+						obj[this.columnNames[j].name] = record[j];
+					}
+					return obj;
+				});
+				$('#spinner').hide();
+				$('#grid').show();
+				return processedData;
+			})
+			.catch(() => {
+				throw ('Failed to fetch records');
```

Comment:
```suggestion
			.catch(error => {
				throw ('Failed to fetch records: ', error);
```
I mean, wouldn't you like to know the error? We know if failed but why?